pkg/aws/ec2/ec2.go

```diff
@@ -217,7 +289,14 @@ func parseENI(iface *ec2_types.NetworkInterface, vpcs ipamTypes.VirtualNetworkMa
 func (c *Client) GetInstances(ctx context.Context, vpcs ipamTypes.VirtualNetworkMap, subnets ipamTypes.SubnetMap) (*ipamTypes.InstanceMap, error) {
 	instances := ipamTypes.NewInstanceMap()
 
-	networkInterfaces, err := c.describeNetworkInterfaces(ctx, subnets)
+	var networkInterfaces []ec2_types.NetworkInterface
+	var err error
+
+	if len(c.instancesFilters) > 0 {
```

Comment:
So this preserves the existing subnet tag filter behavior if no instance filter is set. That is great, because it won't surprise anyone who upgrades with a new behavior.

But I do wonder if we maybe should just always call into `describeNetworkInterfacesFromInstances` here to avoid the additional code complexity. The problem with the subnet tag filter is that it only works here if the subnet tag filter matches the subnet of default ENI on the fetched instances, which is a bit of a surprising behavior. Therefore I do wonder if we should just get rid of the surprising behavior here, and ask users to also set a instance filter if they are concerned about scalability of the query.

It would however require large-scale user to define a instance tag filter before upgrading to avoid overly large result sets, so not sure if that's considered a breaking change. What are the opinions on this? (e.g. @christarazi?)

Reply:
Yeah, I'm curious about your (Cilium maintainers) view. 
I built up the PR this way to avoid any breaking change. 
The truth is, `instance tag filter` will work better than `subnet tag filter` in any complex scenario. My take is that we should deprecate the `subnet tag filter` in a coming version...
That said, we could also bring both filters together in the same function and do the check there...

Reply:
@gandro It sounds like a breaking change to me, since users are definitely configuring Cilium specifically around this. 

@prune998 Let's add an entry to the upgrade docs and have docs around this in general to help communicate what's being done in this PR and what's recommended moving forward.

Reply:
I thought about this some more, and I strongly believe now we should change the behavior of the `subnet tag filter` and not apply it in `GetInstances`. We've now had multiple users reporting it breaking their cluster due to it's behavior being very surprising. Therefore I think changing the behavior (while technically a breaking change) will actually make the life of most users easier.

Therefore, I'm proposing to always call `describeNetworkInterfacesFromInstances` here and delete the old `describeNetworkInterfaces` function. Users who have scalability concerns can use the new `instance tag filter`.

Reply:
@gandro, while I agree with you, do we have to do this now in this PR or do you want to wait and deprecate the `subnet tag filter` later on ?
We could add the new `instance tag filter` now and document deprecation of the`subnet tag filter`, and remove it on the following release ? 
Let me know and I'll update the code.

Reply:
I rebased, corrected, and added the new option into the Helm chart + the upgrade docs as a new parameter ; I also added a deprecation warning for the `subnet tag filter`.
Let me know what you think.

Reply:
`subnet` and `instance` tag filters are just limiting what information Cilium is managing. 
You actually want Cilium to know about ALL the ENIs of the instances that you're using in your cluster. This is what you get by using `instance tag filter`.
This is not guaranty if you use `subnet tag filter`. 
But I don't see any case where you would want to use both.

Cilium is using the value from `cilium-cni-configuration` ConfigMap to pick the subnet to use to add new IPs, not the data filtered by the `subnet tag filter` (as far as I know)
```
          "eni": {
            "subnet-tags":{
              "Usage":"pods"
            }
```

You're right that `subnet tag filter` is useful to limit the number of subnets Cilium will sync, but it is not relevant to use is to `GetInstances`

Reply:
>  You actually want Cilium to know about ALL the ENIs of the instances that you're using in your cluster. This is what you get by using instance tag filter.

But will your current implementation contain the subnet information extracted here:
https://github.com/cilium/cilium/blob/7c1bb3592c46e6d3bc04d3eb8ff37aaf70f06dc3/pkg/aws/ec2/ec2.go#L193-L198

If not, we still will have to fetch the related subnets manually.

Reply:
`subnets` is coming from the `GetSubnets()` func, so it will be either all subnets of the VPC if there's no subnet filter, and some subnets if you use a filter.
My take here, is that the `subnet` filter was implemented as a quick way to limit the number of ENIs (and instances) to maintain. Usually, you have a lot of instances and a lot of ENIs/IPs, but not that many subnets in the same account. And even if it was a hundred subnets, i'm not sure it would have a huge impact.

Ideally, if a user wants to perfectly tune the AWS API search so Cilium would only maintain something that can be used, then they would define both a `subnet` filter so `describeSubnets()` would only grab useful subnets AND an `instance` filter to only grab instances of the cluster.
Of course, we could replace all that and get the list of subnets from the selected instances... but that would be a lot of changes, and I don't know how other providers (alibaba, google, azure) would behave....
 
That said, the outcome is that we should keep the `subnet` filter for `GetSubnets()`  and keep the `instance` filter for `GetInstances()`. And it will be up to the user to ensure the filters matches...

Reply:
> That said, the outcome is that we should keep the subnet filter for GetSubnets() and keep the instance filter for GetInstances(). And it will be up to the user to ensure the filters matches...

Yes, I think that's a reasonable outcome. So to summarize this means we won't deprecate or change the behavior of the `subnet tag filter` option, but make sure we document it properly.

I already started working on a docs PR in https://github.com/cilium/cilium/pull/19276, but it's intentionally kept separate from this PR so we can backport it to older versions.